### ISSUE TITLE
How do I train an HMM on a set of sequences with different lengths?

### ISSUE THREAD

**Issue by lucasmiranda42** [tensorflow/probability#1373]
Hi! I'm trying to train an HMM on data coming from a dataset in which the sequence lengths are variable, which seems hard due to the required num_steps parameter on tfp.distributions.HiddenMarkovModel.

I tried padding and masking using tf.keras.layers.Masking, but the results do not seem correct (the log_prob seems to be affected by the padded values, which is not the desired behaviour). What would be the best way to proceed?

Thank you!

My code, where X_train is the output of passing my padded, initial input through tf.keras.layers.Masking():

    model = tfd.HiddenMarkovModel(
        initial_distribution=initial_distribution,
        transition_distribution=transition_distribution,
        observation_distribution=emission_distribution,
        num_steps=X_train.shape[1],
    )

    optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)
    trainable_variables = [trainable_transitions, trainable_means, trainable_scales]

    # Create a shuffled and batched dataset out of X_train
    X_train = (
        tf.data.Dataset.from_tensor_slices(X_train)
        .shuffle(batch_size)
        .batch(batch_size, drop_remainder=False)
    )

    # Define loss function
    def log_prob(current_batch, current_hmm):
        """Loss function for HMMs in TensorFlow"""

        return tf.reduce_mean(current_hmm.log_prob(current_batch))

    @tf.function(autograph=True)
    def train_op(current_batch, current_hmm, trainable, opt):
        """Training step for HMMs in TensorFlow"""

        with tf.GradientTape() as tape:
            neg_log_prob = -log_prob(current_batch, current_hmm)

        grads = tape.gradient(neg_log_prob, trainable)

        opt.apply_gradients(zip(grads, trainable))

        return neg_log_prob

    for epoch in range(epochs):

        for batch in X_train:
            loss = train_op(batch, model, trainable_variables, optimizer)

**Comment by davmre** [comment]
I think you can achieve this by using `tfd.Masked` to limit the HMM's observation distribution to the steps you want to model (note this requires the `time_varying_observation_distribution=True` flag to your HMM). I wrote up an example here:

https://colab.research.google.com/drive/1wnq7SXX3aJYUFq8RJkvaEW83Asffbo6P?usp=sharing

This same approach works more generally for handling *any* pattern of missing observations in an HMM; they don't necessarily have to be at the end of the series. We'd be very open to contributions to make this more straightforward (for example, HMM methods could take a `mask` argument similar to `tfd.LinearGaussianStateSpaceModel`).

**Comment by lucasmiranda42** [comment]
This is great, thank you! It does work, and I'll try to gather some time and contribute to the implementation as you suggested.
However, when I try to train a distribution using a batched dataset (and therefore a batched mask) I end up having to instantiate the model again for each batch (which is terribly slow). Is there any workaround to this?

Thanks again!
Lucas

**Comment by davmre** [comment]
A good general principle of optimizing TF code is to use `@tf.function` to avoid Python overhead, meaning any time spent doing stuff other than your actual computation, like (re)building models or other complicated objects.

A corollary of this is that TFP treats object construction as 'cheap', and often expects you to do a lot of it, on the assumption that none of this will show up in the `tf.function` trace of the Tensor computations done by your code. TFP objects (distributions, bijectors, etc.) commit to not do any Tensor ops in their constructors, so re-initializing them is 'free' in that sense, even though (as you're seeing) it can be expensive in pure Eager execution.

The main restriction on `tf.function`s is that they take and return Tensors. Here, I think you could probably wrap your training step:

```python
@tf.function(autograph=False)
def do_train_step(batch, mask):
  model = tfd.HiddenMarkovModel(
      initial_distribution=initial_distribution,
      transition_distribution=transition_distribution,
      observation_distribution=tfd.Masked(
          emission_distribution,
          validity_mask=mask[..., tf.newaxis],
      ),
      time_varying_observation_distribution=True,
      num_steps=batch.shape[1],
      validate_args=True,
  )
  return train_op(batch, model, trainable_variables, optimizer)
```

and then run the wrapped step inside the loop:

```python
for epoch in range(epochs):
    for sample in X_train:
        batch, mask = sample
        loss = do_train_step(batch, mask)
```

That will instantiate `HiddenMarkovModel` only once when the training step is first traced, so all subsequent steps should be faster. You might also try playing with the `jit_compile` argument to `tf.function`, which adds some more initial overhead but can sometimes give large speedups.

**Comment by lucasmiranda42** [comment]
I should have thought of that... thanks! It works like a charm :)
